core/src/main/java/hudson/model/ItemGroupMixIn.java

```diff
@@ -329,6 +337,40 @@ public synchronized TopLevelItem createProject( TopLevelItemDescriptor type, Str
     }
 
     /**
+     * Populate a {$link Categories} from a specific {$link ItemGroup}.
+     *
+     * @return
+     */
+    public static Categories getCategories(Authentication a, ItemGroup c) {
+        Categories categories = new Categories();
+        for (TopLevelItemDescriptor descriptor : Items.all(a, c)) {
+            ItemCategory ic = ItemCategoryConfigurator.getCategory(descriptor);
+            int i = 0;
+            boolean found = false;
+            while (i < categories.getItems().size() && !found) {
+                if (categories.getItems().get(i).getId() == ic.getId()) {
+                    Map<String, String> metadata = new HashMap<String, String>();
+                    metadata.put("class", descriptor.clazz.getName());
+                    metadata.put("iconClassName", "item-icon-" + descriptor.clazz.getName().substring(descriptor.clazz.getName().lastIndexOf(".") + 1).toLowerCase());
+                    categories.getItems().get(i).getItems().add(metadata);
+                    found = true;
+                }
+                i++;
+            }
+            if (!found) {
+                Map<String, String> metadata = new HashMap<String, String>();
+                metadata.put("class", descriptor.clazz.getName());
+                metadata.put("iconClassName", "item-icon-" + descriptor.clazz.getName().substring(descriptor.clazz.getName().lastIndexOf(".") + 1).toLowerCase());
+                List<Map<String, String>> temp = new ArrayList<Map<String, String>>();
+                temp.add(metadata);
+                categories.getItems().add(new Category(ic.getId(), ic.getDisplayName(), ic.getDescription(),
+                        ic.getIconClassName(), ic.getWeight(), temp));
+            }
+        }
+        return categories;
+    }
```

Comment:
I think you missed the point I was making about the structure of `getCategories` ala https://github.com/tfennelly/jenkins/commit/b593c44686b7b9ec3bc37faf44a3d2a4e5b57e0b

The following seems a lot easier to understand Vs `while` looks, `found` flags etc

```java
    public static Categories getCategories(Authentication a, ItemGroup c) {
        Categories categories = new Categories();
        for (TopLevelItemDescriptor descriptor : Items.all(a, c)) {
            ItemCategory ic = ItemCategoryConfigurator.getCategory(descriptor);
            Map<String, String> metadata = new HashMap<String, String>();

            metadata.put("class", descriptor.clazz.getName());
            metadata.put("iconClassName", "item-icon-" + descriptor.clazz.getName().substring(descriptor.clazz.getName().lastIndexOf(".") + 1).toLowerCase());

            Category category = categories.getItem(ic.getId());
            if (category != null) {
                category.getItems().add(metadata);
            } else {
                List<Map<String, String>> temp = new ArrayList<Map<String, String>>();
                temp.add(metadata);
                category = new Category(ic.getId(), ic.getDisplayName(), ic.getDescription(),
                        ic.getIconClassName(), ic.getWeight(), temp);
                categories.getItems().add(category);
            }
        }
        return categories;
    }
```